tests/test_functional.py

```diff
@@ -681,3 +683,183 @@ class C(object):
             "2021-06-01.  Please use `eq` and `order` instead."
             == w.message.args[0]
         )
+
+
+class TestSetAttr(object):
+    def test_change(self):
+        """
+        The return value of a hook overwrites the value. But they are not run
+        on __init__.
+        """
+
+        def hook(*a, **kw):
+            return "hooked!"
+
+        @attr.s
+        class Hooked(object):
+            x = attr.ib(on_setattr=hook)
+            y = attr.ib()
+
+        h = Hooked("x", "y")
+
+        assert "x" == h.x
+        assert "y" == h.y
+
+        h.x = "xxx"
+        h.y = "yyy"
+
+        assert "yyy" == h.y
+        assert "hooked!" == h.x
+
+    def test_frozen_attribute(self):
+        """
+        Frozen attributes raise FrozenAttributeError, others are not affected.
+        """
+
+        @attr.s
+        class PartiallyFrozen(object):
+            x = attr.ib(on_setattr=setters.frozen)
+            y = attr.ib()
+
+        pf = PartiallyFrozen("x", "y")
+
+        pf.y = "yyy"
+
+        assert "yyy" == pf.y
+
+        with pytest.raises(FrozenAttributeError):
+            pf.x = "xxx"
```

Comment:
Just FYI.  Mypy will not catch this as an error.  Not important for this diff but something to consider.

Reply:
Oh you mean mypy should be able to catch the presence of a frozen hook? That sounds tough but would be cool indeed.